feat(memorystore): added caching guide (#1)
invertase/java-docs-samples: memorystore/valkey/caching/GUIDE.md

3. **Uses Spring Boot** to expose REST endpoints, providing a simple interface for reading, writing, and invalidating cached data.
4. **Can be Dockerized and deployed to Google Cloud Platform (GCP)** for production, leveraging services like Cloud Run, Cloud SQL, and Memorystore.

By following this guide, you’ll have a reference architecture ready to adapt, test, and deploy to meet the performance needs of your application.

## Architecture Overview

- **Spring Boot Application:** Serves as the middle tier for responding to API calls. When a request is received, the API checks Valkey for cached results; if no entries are found, then the API will retrieve data from the PostgreSQL database and update the cache.
- **Valkey (In-Memory Cache):** A Redis-like memory store that keeps hot data ready to be served instantly.
- **PostgreSQL Database:** Your source of truth for all data. The cache reduces how often the app queries this database.
- **Google Cloud Infrastructure:** Deployed using Terraform, you can host the application on Cloud Run, store data in Cloud SQL for PostgreSQL, and leverage Memorystore for Valkey.

## Steps to Build

1. **Set Up Your Environment:**

>    - Install Java 17 (or later), Maven, Docker, and Docker Compose locally.

Author: ```suggestion
     Install the following:

   - [Docker](https://www.docker.com)
   - [Docker Compose](https://docs.docker.com/compose/)
```

- **User Experience:** Faster responses translate into happier users, which is especially crucial for high-traffic or interactive applications such as gaming and e-commerce platforms.

**Common Use Cases:**

- **E-Commerce:** Quickly serve product catalogs, pricing details, and inventory counts.
- **Gaming:** Cache player stats, sessions, and game states for real-time gaming experiences.
- **API Gateways & Microservices:** Reduce downstream service calls by serving cached API responses.

## What You’ll Build

You’ll set up a caching service that:

1. **Works with a PostgreSQL database** to store long-lived, persistent records.
2. **Incorporates Valkey** as a high-speed, in-memory cache, fronting the PostgreSQL database.
3. **Uses Spring Boot** to expose REST endpoints, providing a simple interface for reading, writing, and invalidating cached data.
4. **Can be Dockerized and deployed to Google Cloud Platform (GCP)** for production, leveraging services like Cloud Run, Cloud SQL, and Memorystore.

By following this guide, you’ll have a reference architecture ready to adapt, test, and deploy to meet the performance needs of your application.

## Architecture Overview

> 

Author: ```suggestion
```

- **Valkey (In-Memory Cache):** A Redis-like memory store that keeps hot data ready to be served instantly.
- **PostgreSQL Database:** Your source of truth for all data. The cache reduces how often the app queries this database.
- **Google Cloud Infrastructure:** Deployed using Terraform, you can host the application on Cloud Run, store data in Cloud SQL for PostgreSQL, and leverage Memorystore for Valkey.

## Steps to Build

1. **Set Up Your Environment:**

     Install the following:

   - [Docker](https://www.docker.com)
   - [Docker Compose](https://docs.docker.com/compose/)
   - Set up a GCP project, enabling Cloud Run, Memorystore, and Cloud SQL APIs if you plan to deploy to Google Cloud.

2. **Download the Example Code:**
   Instead of writing all the code from scratch, we’ve prepared a working demo repository that you can clone and explore. This contains everything you need including Spring Boot configuration, caching logic, Dockerfiles, and Terraform scripts for deployment.

   **Get the code here:**  

>    [https://github.com/invertase/valkey-demos/tree/main/demos/cache](https://github.com/invertase/valkey-demos/tree/main/demos/cache)

Author: ```suggestion
   [https://github.com/GoogleCloudPlatform/java-docs-samples/tree/main/memorystore/valkey/cache](https://github.com/GoogleCloudPlatform/java-docs-samples/tree/main/memorystore/valkey/cache)
```

   Visit the repository’s Terraform directory for detailed instructions and run:

   ```bash
   terraform init
   terraform apply
   ```

   After Terraform has deployed your infrastructure, you’ll have a production-ready caching architecture running in the cloud.

## Scaling and Optimization

As traffic increases, the architecture can scale horizontally:

- **Cloud Run** can automatically scale instances based on load.
- **Memorystore (Valkey)** can be sized or upgraded to handle more cached data or higher throughput.
- **Cloud SQL** can scale vertically or horizontally (with read replicas) as needed.

Fine-tune cache expiration strategies (TTL values) and eviction policies, depending on your data access patterns.

## Conclusion

By combining an in-memory store (Valkey) with a reliable database (PostgreSQL), all orchestrated by a Spring Boot application, you’ve built a caching solution that delivers high performance, reduces database load, and ensures an excellent user experience. Running it in Google Cloud extends these benefits further, providing managed services and easy scaling.

> Ready to dive into the code and start experimenting? Check out the repository for the full project details and follow the instructions to get started:

Author: ```suggestion
For more information check out the repository for the full project details and follow the instructions to get started:
```

> [https://github.com/invertase/valkey-demos/tree/main/demos/cache](https://github.com/invertase/valkey-demos/tree/main/demos/cache)

Author: ```suggestion
[https://github.com/GoogleCloudPlatform/java-docs-samples/tree/main/memorystore/valkey/caching](https://github.com/GoogleCloudPlatform/java-docs-samples/tree/main/memorystore/valkey/caching)
```

# Build a Caching Service on Google Cloud using Valkey, Spring Boot, and PostgreSQL
In this tutorial, we’ll walk through the architectural concepts and deployment steps for creating a high-performance caching service on Google Cloud. Using a combination of Java, Spring Boot, PostgreSQL, and [Valkey](https://github.com/invertase/valkey-demos), an in-memory key-value store with Redis-like interfaces, you can significantly reduce latency, offload your primary database, and deliver consistent, low-latency responses.

## Why Caching Matters

> - **Speed & Latency:** Storing frequently requested data in memory avoids repeated round-trip queries to databases, slashing response times.

Author: ```suggestion
- **Speed & Latency:** Storing frequently requested data in memory avoids repeated round-trip queries to databases, whilst reducing response times.
```

> - **Scalability:** By easing the load on your primary database, caching allows you to handle more requests smoothly as traffic grows.

Author: ```suggestion
- **Scalability:** By reducing the workload on your database, applications can serve data directly from memory, increasing the capacity for requests.
```

> - **User Experience:** Faster responses translate into happier users, which is especially crucial for high-traffic or interactive applications such as gaming and e-commerce platforms.

Author: ```suggestion
```

> **Common Use Cases:**
> 
> - **E-Commerce:** Quickly serve product catalogs, pricing details, and inventory counts.
> - **Gaming:** Cache player stats, sessions, and game states for real-time gaming experiences.
> - **API Gateways & Microservices:** Reduce downstream service calls by serving cached API responses.

Author: ```suggestion
```

## What You’ll Build

You’ll set up a caching service that:

1. **Works with a PostgreSQL database** to store long-lived, persistent records.
2. **Incorporates Valkey** as a high-speed, in-memory cache, fronting the PostgreSQL database.
3. **Uses Spring Boot** to expose REST endpoints, providing a simple interface for reading, writing, and invalidating cached data.
4. **Can be Dockerized and deployed to Google Cloud Platform (GCP)** for production, leveraging services like Cloud Run, Cloud SQL, and Memorystore.

By following this guide, you’ll have a reference architecture ready to adapt, test, and deploy to meet the performance needs of your application.

## Architecture Overview

> The high-level architecture looks like this:

Author: ```suggestion
```

   - Set up a GCP project, enabling Cloud Run, Memorystore, and Cloud SQL APIs if you plan to deploy to Google Cloud.

2. **Download the Example Code:**
   Instead of writing all the code from scratch, we’ve prepared a working demo repository that you can clone and explore. It includes everything you need—Spring Boot configuration, caching logic, Dockerfiles, and Terraform scripts for deployment.

   **Get the code here:**  
   [https://github.com/invertase/valkey-demos/tree/main/demos/cache](https://github.com/invertase/valkey-demos/tree/main/demos/cache)

3. **Review the Code Structure:**
   In the repository, you’ll find:

   - **Spring Boot project files:** Main application class, REST controllers, repository classes, and caching logic.
   - **Configuration files:** For JDBC (PostgreSQL) and Jedis (Valkey) clients, allowing flexible environment-based configurations.
   - **Terraform scripts (optional):** Infrastructure as Code templates to spin up resources on Google Cloud.

   The code is well-organized, with clear separation of concerns:

   - **Application Layer:** REST endpoints expose CRUD operations for cached data.

>    - **Data Access Layer:** Repositories interact with PostgreSQL for persistent storage.

Reviewer: ```suggestion
   - **Data Access Layer:** This contains the logic for interacting with the cache and the PostgreSQL database.
```

   - **Caching Layer:** A controller checks Valkey first and falls back to the database if needed.

4. **Customize the Configuration:**

   - Set your database URL, username, and password via environment variables or configuration files.
   - Point your VALKEY_HOST and VALKEY_PORT environment variables to your Valkey instance (in development, set it to localhost when running via Docker Compose).

5. **Run Locally with Docker Compose:**
   Use Docker Compose to start PostgreSQL, Valkey, and your Spring Boot app together. You’ll have a fully functional local environment that demonstrates how caching accelerates data retrieval.

   Simply run:

   ```bash
   docker compose up --build
   ```

   Once started, you can make `GET`, `POST`, and `DELETE` requests to the REST endpoints to store, retrieve, and invalidate cached data. Check the repository’s README for example commands and endpoints.

6. **Deploying to GCP (Optional):**

>    With Terraform, you can provision:

Author: ```suggestion
   
   ## Install [Terraform](https://developer.hashicorp.com/terraform/tutorials/gcp-get-started)
  Terraform is an Infrastructure as Code (IaC) tool that allows you to define and provision cloud resources in a repeatable, automated way.

   With Terraform, you can provision:
```

4. **Can be Dockerized and deployed to Google Cloud Platform (GCP)** for production, leveraging services like Cloud Run, Cloud SQL, and Memorystore.

By following this guide, you’ll have a reference architecture ready to adapt, test, and deploy to meet the performance needs of your application.

## Architecture Overview

- **Spring Boot Application:** Serves as the middle tier for responding to API calls. When a request is received, the API checks Valkey for cached results; if no entries are found, then the API will retrieve data from the PostgreSQL database and update the cache.
- **Valkey (In-Memory Cache):** A Redis-like memory store that keeps hot data ready to be served instantly.
- **PostgreSQL Database:** Your source of truth for all data. The cache reduces how often the app queries this database.
- **Google Cloud Infrastructure:** Deployed using Terraform, you can host the application on Cloud Run, store data in Cloud SQL for PostgreSQL, and leverage Memorystore for Valkey.

## Steps to Build

1. **Set Up Your Environment:**

     Install the following:

   - [Docker](https://www.docker.com)
   - [Docker Compose](https://docs.docker.com/compose/)

>    - Set up a GCP project, enabling Cloud Run, Memorystore, and Cloud SQL APIs if you plan to deploy to Google Cloud.

Author: ```suggestion

     Project Setup

   - Set up a GCP project
   - Enable Cloud Run API
   - Enable Memorystore API
   - Cloud SQL APIs
```

> 2. **Download the Example Code:**
>    Instead of writing all the code from scratch, we’ve prepared a working demo repository that you can clone and explore. It includes everything you need—Spring Boot configuration, caching logic, Dockerfiles, and Terraform scripts for deployment.

Author: ```suggestion
2. **Download the Example Code:**
   Instead of writing all the code from scratch, we’ve prepared a working demo repository that you can clone and explore. This contains everything you need including Spring Boot configuration, caching logic, Dockerfiles, and Terraform scripts for deployment.
```

   **Get the code here:**  
   [https://github.com/invertase/valkey-demos/tree/main/demos/cache](https://github.com/invertase/valkey-demos/tree/main/demos/cache)

3. **Setup the application:**
   In the repository, you’ll find:

   - **Spring Boot project files:** Main application class, REST controllers, repository classes, and caching logic.
   - **Configuration files:** For JDBC (PostgreSQL) and Jedis (Valkey) clients, allowing flexible environment-based configurations.
   - **Terraform scripts (optional):** Infrastructure as Code templates to spin up resources on Google Cloud.

>    The code is well-organized, with clear separation of concerns:

Author: ```suggestion
```

# Build a Caching Service on Google Cloud using Valkey, Spring Boot, and PostgreSQL

> 
> 

Author: ```suggestion
```

> **Build a Caching Service on Google Cloud using Valkey, Spring Boot, and PostgreSQL**

Author: ```suggestion
# Build a Caching Service on Google Cloud using Valkey, Spring Boot, and PostgreSQL
```

Reviewer 2: `# Build a Caching Service on Google Cloud using Valkey, Spring Boot, and PostgreSQL`

> Modern applications need to deliver fast, responsive user experiences at scale. Whether you're hosting an e-commerce storefront, running a real-time gaming backend, or providing dynamic content APIs, performance and reliability are key. One tried-and-true way to meet these demands is by adding a caching layer.

Author: ```suggestion
```

> Modern applications need to deliver fast, responsive user experiences at scale. Whether you're hosting an e-commerce storefront, running a real-time gaming backend, or providing dynamic content APIs, performance and reliability are key. One tried-and-true way to meet these demands is by adding a caching layer.

Reviewer: fast and responsive are pretty similar, maybe something like:

"Modern applications demand performance and scalability to provide seamless user experiences"

> In this tutorial, we’ll walk through the architectural concepts and deployment steps for creating a high-performance caching service on Google Cloud. Using a combination of Java, Spring Boot, PostgreSQL, and [Valkey](https://github.com/invertase/valkey-demos), an in-memory key-value store with Redis-like interfaces, you can significantly reduce latency, offload your primary database, and deliver consistent, low-latency responses.

Author: ```suggestion
Modern applications need to deliver fast, responsive user experiences at scale.

In this tutorial, we’ll walk through the architectural concepts and deployment steps for creating a high-performance caching service on Google Cloud. Using a combination of Java, Spring Boot, PostgreSQL, and [Valkey](https://github.com/invertase/valkey-demos) you can significantly reduce latency, while offload ing your primary database

TODO

```

> In this tutorial, we’ll walk through the architectural concepts and deployment steps for creating a high-performance caching service on Google Cloud. Using a combination of Java, Spring Boot, PostgreSQL, and [Valkey](https://github.com/invertase/valkey-demos), an in-memory key-value store with Redis-like interfaces, you can significantly reduce latency, offload your primary database, and deliver consistent, low-latency responses.

Reviewer: (update repo links throughout)

## Why Caching Matters

- **Speed & Latency:** Storing frequently requested data in memory avoids repeated round-trip queries to databases, slashing response times.
- **Scalability:** By easing the load on your primary database, caching allows you to handle more requests smoothly as traffic grows.

> - **User Experience:** Faster responses translate into happier users, which is especially crucial for high-traffic or interactive applications such as gaming and e-commerce platforms.

Author: Cost savings maybe? Is the user experience improvement inferred from the first two points.

Reviewer 2: What's your thinking behind the cost saving, because I can't identify how it would be cost saving? If anything, it'd cost them more, because they need to pay for the Valkey instances. However, the improvement in performance could increase things like SEO, conversion rates, and lower bounce rates. The user experience bullet kinda conveys that already though.

> **Common Use Cases:**

Author: Not sure if we need this?

- **E-Commerce:** Quickly serve product catalogs, pricing details, and inventory counts.
- **Gaming:** Cache player stats, sessions, and game states for real-time gaming experiences.
- **API Gateways & Microservices:** Reduce downstream service calls by serving cached API responses.

## What You’ll Build

You’ll set up a caching service that:

1. **Works with a PostgreSQL database** to store long-lived, persistent records.
2. **Incorporates Valkey** as a high-speed, in-memory cache, fronting the PostgreSQL database.
3. **Uses Spring Boot** to expose REST endpoints, providing a simple interface for reading, writing, and invalidating cached data.

> 4. **Can be Dockerized and deployed to Google Cloud Platform (GCP)** for production, leveraging services like Cloud Run, Cloud SQL, and Memorystore.

Author: ```suggestion
4. **Can be Dockerized and deployed to Google Cloud Platform (GCP)** for production, allowing the bundled deployment of  services like Cloud Run, Cloud SQL, and Memorystore.
```

By following this guide, you’ll have a reference architecture ready to adapt, test, and deploy to meet the performance needs of your application.

## Architecture Overview

The high-level architecture looks like this:

> - **Spring Boot Application:** Serves as the middle tier, responding to API calls. It checks Valkey first for cached results; if not found, it retrieves data from PostgreSQL, then updates the cache.

Author: ```suggestion
- **Spring Boot Application:** Serves as the middle tier for responding to API calls. When a request is received, the API checks Valkey for cached results; if no entries are found, then the API will retrieve data from the PostgreSQL database and update the cache.
```

- **Valkey (In-Memory Cache):** A Redis-like memory store that keeps hot data ready to be served instantly.
- **PostgreSQL Database:** Your source of truth for all data. The cache reduces how often the app queries this database.
- **Google Cloud Infrastructure:** Deployed using Terraform, you can host the application on Cloud Run, store data in Cloud SQL for PostgreSQL, and leverage Memorystore for Valkey.

## Steps to Build

1. **Set Up Your Environment:**

   - Install Java 17 (or later), Maven, Docker, and Docker Compose locally.
   - Set up a GCP project, enabling Cloud Run, Memorystore, and Cloud SQL APIs if you plan to deploy to Google Cloud.

2. **Download the Example Code:**

>    Instead of writing all the code from scratch, we’ve prepared a working demo repository that you can clone and explore. It includes everything you need—Spring Boot configuration, caching logic, Dockerfiles, and Terraform scripts for deployment.

Reviewer 2: Although not necessarily proof of AI use, AI does love to use em-dashes all of the time. [Other people recognise this too](https://www.reddit.com/r/ChatGPT/search/?q=em+dashes&cId=2b9e3a07-2dd3-4942-8e8a-4fe379ae0306&iId=9959bc31-be8f-487f-9254-08543c9d44c1).

Is it worth re-writing this to replace the em-dash?

Author: Yes, try to re-write this as you wold through your own natural language.

   **Get the code here:**  

>    [https://github.com/invertase/valkey-demos/tree/main/demos/cache](https://github.com/invertase/valkey-demos/tree/main/demos/cache)

Reviewer: links will have to be updated to point to the right repo

> 3. **Review the Code Structure:**

Author: ```suggestion
3. **Setup the application:**
```

   In the repository, you’ll find:

   - **Spring Boot project files:** Main application class, REST controllers, repository classes, and caching logic.
   - **Configuration files:** For JDBC (PostgreSQL) and Jedis (Valkey) clients, allowing flexible environment-based configurations.
   - **Terraform scripts (optional):** Infrastructure as Code templates to spin up resources on Google Cloud.

>    The code is well-organized, with clear separation of concerns:

Author: ```suggestion
   The application contains 3 separate layers:
```

>    - **Application Layer:** REST endpoints expose CRUD operations for cached data.

Author: ```suggestion
   - **Application Layer:** REST API endpoints expose CRUD operations for fetching, updating and deleting data.
```

>    - **Data Access Layer:** Repositories interact with PostgreSQL for persistent storage.

Author: ```suggestion
   - **Data Access Layer:** This is the main database hosting a PostgreSQL database for persistent storage.
```

   - **Caching Layer:** A controller checks Valkey first and falls back to the database if needed.

4. **Customize the Configuration:**

   - Set your database URL, username, and password via environment variables or configuration files.
   - Point your VALKEY_HOST and VALKEY_PORT environment variables to your Valkey instance (in development, set it to localhost when running via Docker Compose).

5. **Run Locally with Docker Compose:**
   Use Docker Compose to start PostgreSQL, Valkey, and your Spring Boot app together. You’ll have a fully functional local environment that demonstrates how caching accelerates data retrieval.

   Simply run:

   ```bash

>    docker-compose up --build

Reviewer: ```suggestion
   docker compose up --build
```

Reviewer 2: [Good catch](https://docs.docker.com/compose/intro/compose-application-model/)

   ```

>    Once started, you can make GET, POST, and DELETE requests to the REST endpoints to store, retrieve, and invalidate cached data. Check the repository’s README for example commands and endpoints.

Reviewer: ```suggestion
   Once started, you can make `GET`, `POST`, and `DELETE` requests to the REST endpoints to store, retrieve, and invalidate cached data. Check the repository’s README for example commands and endpoints.
```

maybe? not sure.

Reviewer 2: I like this

6. **Deploying to GCP (Optional):**

>    With Terraform, you can provision:

Reviewer: might be worth a short sentence somewhere (not sure where) defining terms like terraform and valkey, e.g

```
Terraform is an Infrastructure as Code (IaC) tool that allows you to define and provision cloud resources in a repeatable, automated way.
```

Reviewer 2: Line 35 has a small definition of Valkey. Terraform could perhaps use something similar. Line 37 could define Terraform?

   - **Cloud Run:** Runs your containerized Spring Boot application in a fully managed, serverless environment.
   - **Cloud SQL for PostgreSQL:** Provides a fully managed PostgreSQL instance.
   - **Memorystore (Valkey):** A fully managed Redis-compatible cache.

>    Adjust the Terraform variables for your project, initialize Terraform, and apply the configuration. Once deployed, set up environment variables in Cloud Run to point to the appropriate instances.

Author: Which variables, what does the user need to setup?

   Visit the repository’s Terraform directory for detailed instructions and run:

   ```bash
   terraform init
   terraform apply
   ```

>    After Terraform finishes, you’ll have a production-ready caching architecture running in the cloud.

Author: ```suggestion
   After Terraform has deployed your infrastructure, you’ll have a production-ready caching architecture running in the cloud.
```

> ## Testing and Validation

Author: Do we need this?

- **Functional Tests:** Use `curl` or a tool like Postman to send requests to your deployed Cloud Run service. Ensure that reads are fast, and updates reflect in both the cache and database.
- **Load Testing:** Run load tests (e.g., with JMeter or Locust) to confirm performance gains. Compare response times with and without caching enabled.

> - **Monitoring & Logging:** Integrate Google Cloud’s operations suite (formerly Stackdriver) to monitor query times, latency, and error rates, ensuring your caching strategy is delivering the intended benefits.

Author: Do we need this?

## Scaling and Optimization

As traffic increases, the architecture can scale horizontally:

- **Cloud Run** can automatically scale instances based on load.
- **Memorystore (Valkey)** can be sized or upgraded to handle more cached data or higher throughput.
- **Cloud SQL** can scale vertically or horizontally (with read replicas) as needed.

Fine-tune cache expiration strategies (TTL values) and eviction policies, depending on your data access patterns.

## Conclusion

By combining an in-memory store (Valkey) with a reliable database (PostgreSQL), all orchestrated by a Spring Boot application, you’ve built a caching solution that delivers high performance, reduces database load, and ensures an excellent user experience. Running it in Google Cloud extends these benefits further, providing managed services and easy scaling.

> Ready to dive into the code and start experimenting? Check out the repository for the full project details and follow the instructions to get started:

Reviewer: Maybe a more direct call to action, just a suggestion:

Start building your caching service today! Clone the repository, explore the code, and follow the steps to create a scalable, high-performance solution tailored to your application’s needs.

Reviewer 2: Surely if you were to clone the repository, you'd clone the entirety of the Google Docs repo? That probably wouldn't be ideal to the user. Unless there's a way to specifically clone the valkey caching stuff?

Author: Agree this can be improved.

> [https://github.com/invertase/valkey-demos/tree/main/demos/cache](https://github.com/invertase/valkey-demos/tree/main/demos/cache)

Reviewer 2: Needs to point to right repo

> # Build a Caching Service on Google Cloud using Valkey, Spring Boot, and PostgreSQL

Author: ```suggestion
# Building a Caching Service on Google Cloud using Valkey, Spring Boot, and PostgreSQL
```

Modern applications need to deliver fast, responsive user experiences at scale.

> In this tutorial, we’ll walk through the architectural concepts and deployment steps for creating a high-performance caching service on Google Cloud. Using a combination of Java, Spring Boot, PostgreSQL, and [Valkey](https://github.com/invertase/valkey-demos) you can significantly reduce latency, while offload ing your primary database

Author: ```suggestion
In this tutorial, we’ll walk through the architectural concepts and deployment steps for creating a high-performance caching service on Google Cloud. Using a combination of Java, Spring Boot, PostgreSQL, and [Valkey](https://github.com/invertase/valkey-demos) you can significantly reduce latency, while offloading your primary database
```

6. **Deploying to GCP (Optional):**
   
   ## Install [Terraform](https://developer.hashicorp.com/terraform/tutorials/gcp-get-started)
  Terraform is an Infrastructure as Code (IaC) tool that allows you to define and provision cloud resources in a repeatable, automated way.

   With Terraform, you can provision:

   - **Cloud Run:** Runs your containerized Spring Boot application in a fully managed, serverless environment.
   - **Cloud SQL for PostgreSQL:** Provides a fully managed PostgreSQL instance.
   - **Memorystore (Valkey):** A fully managed Redis-compatible cache.

   Adjust the Terraform variables for your project, initialize Terraform, and apply the configuration. Once deployed, set up environment variables in Cloud Run to point to the appropriate instances.

   Visit the repository’s Terraform directory for detailed instructions and run:

   ```bash
   terraform init
   terraform apply
   ```

   After Terraform has deployed your infrastructure, you’ll have a production-ready caching architecture running in the cloud.

> ## Testing and Validation
> 
> - **Functional Tests:** Use `curl` or a tool like Postman to send requests to your deployed Cloud Run service. Ensure that reads are fast, and updates reflect in both the cache and database.
> - **Load Testing:** Run load tests (e.g., with JMeter or Locust) to confirm performance gains. Compare response times with and without caching enabled.
> - **Monitoring & Logging:** Integrate Google Cloud’s operations suite (formerly Stackdriver) to monitor query times, latency, and error rates, ensuring your caching strategy is delivering the intended benefits.

Author: ```suggestion
```
